### PR TITLE
feat(vlm): add Gemma 4 31B joint drafter example config

### PR DESCRIPTION
**Pull request by khazic**
Adds gemma4_31b_joint_drafter_tulu_magicoder_mix.yaml, a 31B analog of the existing 4B joint-drafter example.

Why: the joint-drafter examples only cover Gemma 4 E4B (4B). Google also releases a drafter for the 31B backbone (google/gemma-4-31B-it-assistant), so 31B joint-drafter training works with the stock composite. This adds a ready-to-run config for it.

Notes:
- backbone_hidden_size 5376 and the grouped-query KV-head layout (16 sliding, 4 full) of the released 31B drafter match the 31B base, so no model code changes are needed.
- The drafter shared-KV path is TP and PP unsafe, so tp_size and pp_size stay at 1 and the 31B base is sharded with FSDP2 on a single 8x80GB node, with base activation checkpointing.
- Mirrors the structure and hyperparameters of the 4B text-mix config.

Testing: tools/lint_example_yamls.py passes on the new config; the recipe was validated end-to-end on 8x80GB and the joint loss decreases.

### CODE REVIEW

**Comment by khazic** [comment]
Validated this recipe end-to-end on a single 8x80GB node (Gemma 4 31B base + the released 31B drafter, FSDP2, tp_size 1, base activation checkpointing, K=4, bf16, max_length 2048). The base and drafter load and training is stable; per-GPU memory is flat at ~71 GiB and the joint loss tracks as expected (the drafter term dominates while the base term stays low). Ran cleanly past step 48 with memory unchanged.

At this sequence length the activation churn fragments the allocator, so set `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` to avoid a reserved-but-unallocated OOM near 80 GiB.

Sample steps (reported `loss` = base + 0.01 * drafter):

```
[joint-drafter] L_base=0.0302 L_drafter=1.1223 L_total=0.0414 (lambda=0.010)
step 0  | loss 1.1172 | grad_norm 125.30 | mem 56.76 GiB | tps 40.76(5.10/gpu) | num_label_tokens 863
step 1  | loss 1.3448 | grad_norm 123.24 | mem 71.29 GiB | tps 73.63(9.20/gpu) | num_label_tokens 407
step 5  | loss 1.1189 | grad_norm 222.76 | mem 71.29 GiB | tps 175.19(21.90/gpu) | num_label_tokens 766
step 10 | loss 1.2949 | grad_norm 140.39 | mem 71.29 GiB | tps 60.64(7.58/gpu) | num_label_tokens 393
step 15 | loss 1.3337 | grad_norm 546.34 | mem 71.29 GiB | tps 48.55(6.07/gpu) | num_label_tokens 263
step 20 | loss 0.9541 | grad_norm 96.87  | mem 71.29 GiB | tps 105.87(13.23/gpu) | num_label_tokens 523
step 24 | loss 0.7882 | grad_norm 113.14 | mem 71.29 GiB | tps 114.93(14.37/gpu) | num_label_tokens 807
step 30 | loss 1.2658 | grad_norm 91.08  | mem 71.29 GiB | tps 59.69(7.46/gpu) | num_label_tokens 502
```

**Review by HuiyingLi — APPROVED** [review]
Thank you for validating the recipe.
